toqito/nonlocal_games/binary_constraint_system_game.py

```diff
@@ -0,0 +1,235 @@
+    pairing it with the corresponding parity from ``b``. The bitmask representation
+    can be useful for analyzing linear system games.
+
+    Examples
+    ========
+
+    .. jupyter-execute::
+
+        >>> import numpy as np
+        >>> from toqito.nonlocal_games.binary_constraint_system_game import generate_solution_group
+
+        >>> M = np.array([[1, 1, 0], [0, 1, 1]], dtype=int)
+        >>> b = np.array([0, 1], dtype=int)
+        >>> row_masks, parity = generate_solution_group(M, b)
+
+        >>> print("Row masks:", row_masks)
+        >>> print("Parity:", parity)
+
+    The mehod used to determine the existing of perfect commuting strategy was originally introduced in :cite:`William_2016_Perfect`.
+
+    References
+    ==========
+    .. bibliography::
+        :filter: docname in docnames
+
+
+    :param M : A binary matrix of shape ``(m, n)``.
+    :param b : A binary vector of length ``m``.
+    :return: A list of integer bitmasks (one per row of ``M``).
+    :return: A list of parity values derived from ``b``.
+    
+    """
+    # Ensure M and b are binary (0/1)
+    M = np.array(M, dtype=int) % 2
+    b = np.array(b, dtype=int) % 2
+
+    # Create an array of powers of 2 for each column: [1, 2, 4, ..., 2^(n-1)]
+    powers = 1 << np.arange(M.shape[1])
+    return (M * powers).sum(axis=1).astype(int).tolist(), b.astype(int).tolist()
+
+def _has_undirected_cycle(nodes: list[int], edges: list[tuple[int, int]]) -> bool:
+    r"""Detect whether the undirected graph formed by the given nodes and edges contains any cycle.
+
+    This function builds an adjacency list from the given list of nodes and edges,
+    and then performs a depth-first search (DFS) with parent tracking to detect cycles
+    in the undirected graph. A cycle is detected if the DFS encounters a previously
+    visited node that is not the parent of the current node.
+
+    The function returns True if at least one cycle exists, and False otherwise.
+
+    Examples
+    ==========
+    .. jupyter-execute::
+
+        >>> from binary_constraint_system_game import _has_undirected_cycle
+
+        # Example 1: Graph with a cycle
+        >>> nodes = [0, 1, 2]
+        >>> edges = [(0, 1), (1, 2), (2, 0)]
+        >>> print(_has_undirected_cycle(nodes, edges))  # Output: True
+
+        # Example 2: Graph with no cycle (a tree)
+        >>> nodes = [0, 1, 2]
+        >>> edges = [(0, 1), (1, 2)]
+        >>> print(_has_undirected_cycle(nodes, edges))  # Output: False
+        
+    :param nodes : A list of integer node labels (e.g., [0, 1, 2, ...]).
+    :param edges : A list of undirected edges, each represented as a tuple (u, v) where u and v are nodes from the 'nodes' list.
+    :return: ``True`` if the undirected graph contains at least one cycle, False otherwise.
+
+    Notes
+    ==========
+    - This function is used internally to replace NetworkX's cycle_basis call for
```

Comment:
Do we need to replace this function? Is there a reason not to just use the `cycle_basis` function as provided by `networkx` already?